### PR TITLE
Add GitHub action workflow

### PR DESCRIPTION
**Pull request by mickmister**

### CODE REVIEW

**Comment by mickmister** [comment]
@hanzei Please take a look at this when you have the chance. The reason why the workflow is not running in https://github.com/mattermost/mattermost-app-google-drive/pull/2 is because there are currently no workflows in the master branch of this project. Once this PR is merged, master will fail due to missing scripts (such as `npm run lint`), but they should all pass on https://github.com/mattermost/mattermost-app-google-drive/pull/2 if the code correctly agrees with lint etc.

**Comment by mickmister** [comment]
Instead of having this PR fail CI after merge, we should have a basic package.json that contains the appropriate commands that just echo "TODO". The merge conflicts won't be difficult to resolve

**Comment by hanzei** [comment]
> Instead of having this PR fail CI after merge, we should have a basic package.json that contains the appropriate commands that just echo "TODO". The merge conflicts won't be difficult to resolve

Sounds good to me :+1:

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+name: CI
+
+on:
+  push:
+    branches:
+    - 'ci/**'
+    - 'master'
+    - 'develop'
```
Comment (hanzei) [inline .github/workflows/ci.yml:8]: Why should CI only run on these branches?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  pull_request:
+    branches:
+      - '**'
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
```
Comment (hanzei) [inline .github/workflows/ci.yml:17]: Do we want to control the node version used with something like https://github.com/actions/setup-node?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    - name: Install modules
+      run: npm i
```
Comment (hanzei) [inline .github/workflows/ci.yml:19]: 0/5 use `npm ci` here

**Review by hanzei — CHANGES_REQUESTED** [review]
Neat, thanks for looking into GitHub actions

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mickmister) [reply .github/workflows/ci.yml:8]: I'm not sure how I landed on this decision. For some reason I was thinking that running ci should always be on purpose, but that doesn't make any sense. Running on every branch seems best :+1:

**Review by hanzei — APPROVED** [review]
The code could be a bit more DRY, but it's fine for now.

**Comment by hanzei** [comment]
Do GitHub actions need to be enabled?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - '**'
+  pull_request:
+    branches:
+      - '**'
```
Comment (hanzei) [inline .github/workflows/ci.yml:9]: If CI runs for every push, why does it also need to run per pull request? For contributors?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 4 above)
Comment (mickmister) [reply .github/workflows/ci.yml:9]: @hanzei Yes that's the idea

**Comment by mickmister** [comment]
@hanzei Here's what the output looks like with the config in this PR. Note that it only takes 5 seconds because of the current state of the lint/test commands in `package.json`

image

I don't like how the CI jobs are duplicated here. I remember this is actually why I limited the branches for the `push` event to just `master`, `develop` (Ancient uses a `develop` branch), and `ci/**`.

There doesn't seem to be a "correct" solution to this that results in being able to push any branch and kick off CI https://github.com/orgs/community/discussions/26276. There's an interesting action that claims to solve this. It may be worth looking into https://github.com/marketplace/actions/skip-duplicate-actions#skip-concurrent-workflow-runs

**Comment by mickmister** [comment]
@hanzei I'd like to get this merged so we can apply this here and other repos. I'm 0/5 on what we do in regards to pushes and branches. Maybe we should just do `master`, but I wanted to provide another way, without duplicating the checks in PRs. Maybe it's worth compromise with the duplicated checks, to allow for convenience and flexibility.

Also 3/5, I think we should set up a centralized GitHub action workflow definition in a different repo, and reference it in each App repo. This has already been set up for plugins:

https://github.com/mattermost/mattermost-plugin-gitlab/pull/350
https://github.com/mattermost/actions-workflows/tree/main/.github/workflows

Maybe we should do this the right way on the first iteration of this, and locate these workflows there.

**Comment by mickmister** [comment]
cc @spirosoik in case you have any insight on how those CI workflows can be applied to the Mattermost App repos. It seems like there is a lot more going on in those workflows than just CI checks, and I'm wondering what would be the overlap with the Mattermost App projects.

**Comment by mickmister** [comment]
@hanzei I'd like to have this merged before https://github.com/mattermost/mattermost-app-google-drive/pull/2 is merged, so we can have the CI checks run on that PR

File: .github/workflows/ci.yml (hunk 4 above)
Comment (hanzei) [reply .github/workflows/ci.yml:9]: Maybe I'm misunderstanding how the GitHub CI configuration works.

If a contributor forked the code in `master` and have a CI config with 
```yml
on:
  push:
    branches:
      - '**'
```

why is there a need to also add a trigger for `pull_request`?

**Comment by hanzei** [comment]
:+1: for merging the PR

Maybe we can make it CI run lest often: https://github.com/mattermost/mattermost-app-google-drive/pull/3/files#r1090632331

File: .github/workflows/ci.yml (hunk 4 above)
Comment (mickmister) [reply .github/workflows/ci.yml:9]: @hanzei Right, I think I goofed there. Let's make sure the shared GitHub action config is correct.

Comment (hanzei) [reply .github/workflows/ci.yml:9]: Submitted https://github.com/mattermost/mattermost-app-google-drive/pull/13

**Pull request merged**